Fix/calendar reminders

matthias480 opened the pull request:
* Resolves: #41203

## Summary

Following problems are fixed:

* The Groupware setting "Send reminder notifications to calendar sharees as well" (`sendEventRemindersToSharedUsers`) behaved exactly the other way round.
  If it was activated, sharees din't receive a notification.
* Members of groups with names that contain special characters (also spaces) didn't receive notifications.
* Calendar principal user received reminder notification twice if he/she was also a sharee (emails not affected as EmailProvider filters duplicates)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] ~Screenshots before/after for front-end changes~
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or **is not required**
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

matthias480 committed:
fix(dav): Correctly consider setting 'sendEventRemindersToSharedUsers'

Signed-off-by: Matthias Meusburger <matthias.meusburger@gmx.at>

matthias480 committed:
fix(dav): Decode User/Group IDs to correctly handle special characters (including spaces)

Signed-off-by: Matthias Meusburger <matthias.meusburger@gmx.at>

matthias480 committed:
fix(dav): Do not add principal to reminder recipients twice

Signed-off-by: Matthias Meusburger <matthias.meusburger@gmx.at>

matthias480 commented:
/backport to stable31

matthias480 commented:
/backport to stable30

matthias480 commented:
/backport to stable29

matthias480 commented:
/backport to stable28

matthias480 commented:
/backport to stable27

matthias480 commented:
Hi @SebastianKrupinski

Thank you!

B. OK. I made a fork, because I had no permission to push my branch.
C. OK, I removed 28 and 27.
A. **How to test:**

**Preparation:**

* Have at least 3 users with email addresses.
* Configure your instance to be able to send emails.
* Enable Groupware setting "Send notifications for events".
* Run `php occ config:app:set dav sendEventRemindersMode --value occ` in order to make sending of event reminders triggerable via occ command

**Issue 1:**

1. Create new calendar with user 1.
2. Share calendar with user 2 (write access).
3. Enable Groupware setting "Send reminder notifications to calendar sharees as well".
4. Create new event with one email reminder and one notification reminder.
5. Save the event.
6. Trigger sending when reminders are due via `php occ dav:send-event-reminders`.
7. Observe:
   * Without this PR:
     * User 1 receives reminder mail and notification.
     * User 2 doesn't receive any reminder although "Send reminder notifications to calendar sharees as well" is enabled.
   * With this PR:
     * User 1 and user 2 receive reminder mail and notification.
8. Disable Groupware setting "Send reminder notifications to calendar sharees as well".
9. Repeat 4-6
10. Observe:
    * Without this PR:
      * User 1 receives reminder mail and notification.
      * User 2 receives reminder mail and notification although "Send reminder notifications to calendar sharees as well" is disabled.
    * With this PR:
      * User 1 receives reminder mail and notification.
      * User 2 doesn't receive any reminder.

**Issue 2 + 3:**

1. Create new group with name "Group1".
2. Create new group with name "Group Two".
3. Add user 1 and user 3 to "Group1", add user 2 to "Group Two".
4. Create new calendar with user 1.
5. Share calendar with both groups (write access).
6. Enable Groupware setting "Send reminder notifications to calendar sharees as well". Without this PR: Disable setting since the behavior is the opposite..
7. Create new event with one email reminder and one notification reminder.
8. Save the event.
9. Trigger sending when reminders are due via `php occ dav:send-event-reminders`.
10. Observe:
    * Without this PR:
      * User 1 receives two notifications (because he is the owner/principal and member of Group1) and one mail (mail duplicates are filtered in EmailProvider).
      * User 2 doesn't receive any reminder because group name contains special character (space).
      * User 3 receives reminder mail and notification because group name doesn't contain special characters.
    * With this PR:
      * User 1, 2 and 3 receive one notification and one reminder mail.

Remark: I tested with cron instead of `php occ dav:send-event-reminders`.

SebastianKrupinski commented:
Hi @matthias480 

Thank you for the very thoroughly step by step, it make it easier for us to reproduce the issues and test the fixes.

 Please give me a few days to test this.

matthias480 committed:
feat(dav): Add from name in reminder mail and localize subject

Signed-off-by: Matthias Meusburger <matthias.meusburger@gmx.at>

matthias480 commented:
Hi @SebastianKrupinski 

Not really the preferred way, but since it is in the same file, I added another commit in this PR with two small improvements. I hope, that's OK.

* Added the From name in the reminder email (analogous to other system mails)
* Localized the email's subject (currently it's hard-coded: "Notification:")
  * Actually I would prefer "Reminder:" but I didn't want to touch the current state

SebastianKrupinski commented:
Hi @matthias480 

Thank you, I have not had a chance to test this yet, but will review it test it this week.

niklaswolf commented:
> The Groupware setting "Send reminder notifications to calendar sharees as well" (sendEventRemindersToSharedUsers) behaved exactly the other way round.
If it was activated, sharees din't receive a notification.

I came to the exact same conclusion debugging this on my own, so I confirm the fix for this 👍
